lib/srv/db/redis/connection.go

```diff
@@ -0,0 +1,109 @@
+/*
+
+ Copyright 2022 Gravitational, Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+*/
+
+package redis
+
+import (
+	"net"
+	"net/url"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+// DefaultPort is the Redis default port.
+const DefaultPort = "6379"
+
+const (
+	URIScheme    = "redis"
+	URISchemeSSL = "rediss"
+)
+
+// ConnectionMode defines the mode in which Redis is configured. Currently, supported are single and cluster.
+type ConnectionMode string
+
+const (
+	// Single mode should be used when connecting to a single Redis instance.
+	Single ConnectionMode = "single"
+	// Cluster mode should be used then connecting to a Redis Cluster.
+	Cluster ConnectionMode = "cluster"
+)
+
+// ConnectionOptions defines Redis connection options.
+type ConnectionOptions struct {
+	// mode defines Redis connection mode like cluster or single instance.
+	mode ConnectionMode
+	// address of Redis instance.
+	address string
+	// port on which Redis expects new connections.
+	port string
+}
+
+// ParseRedisURI parses a Redis connection string and returns the parsed
+// connection options like address and connection mode.
+// ex: rediss://redis.example.com:6379?mode=cluster
+func ParseRedisURI(uri string) (*ConnectionOptions, error) {
+	if uri == "" {
+		return nil, trace.BadParameter("Redis uri is empty")
+	}
+
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, trace.BadParameter("failed to parse Redis URI: %v", err)
+	}
+
+	switch u.Scheme {
+	case URIScheme, URISchemeSSL:
+	default:
+		return nil, trace.BadParameter("invalid Redis URI scheme: %q. Expected %q or %q.",
+			u.Scheme, URIScheme, URISchemeSSL)
```

Comment:
In case of mongo db where schema and additional params are also used the URL also allows set url to host:port like `--uri=mongo.example.com:27017 `  where mongo params can be specified by 
`--uri="mongodb://mongo1.example.com:27017,mongo2.example.com:27017/?replicaSet=rs0&readPreference=secondary"`

wonder if we should be consistent in this behavior an allow for just `host:port` in URL

Reply:
I agree that we should be consistent with existing implementations, but there is one thing. that I want to mention:
1. `redis://localhost:6379?mode=cluster` - this is a correct URI
2. `localhost:6379` - this is correct `host:port`
3. `localhost:6379?mode=cluster` this is incorrect URI and it's not `port:host` format.
I'm fine with supporting 1 and 2, but I'd reject the last one, which implies that cluster mode requires to use the first form, and the second is only available for a standalone instances.

@smallinsky Any thoughts?

Reply:
support for 1 and 2 seems to reasonable.